src/inmanta/server/services/compilerservice.py

```diff
@@ -477,27 +478,39 @@ async def is_compiling(self, environment_id: uuid.UUID) -> Apireturn:
             return 200
         return 204
 
-    async def _run(self, compile: data.Compile) -> None:
-        """
-        Runs a compile request. At completion, looks for similar compile requests based on _compile_merge_key and marks
-        those as completed as well.
-        """
-        now = datetime.datetime.now()
-
+    async def _calculate_recompile_wait(self, compile: data.Compile):
+        last_run = await data.Compile.get_last_run(compile.environment)
         wait_time = opt.server_autrecompile_wait.get()
-
-        lastrun = await data.Compile.get_last_run(compile.environment)
-        if not lastrun:
+        if not last_run or wait_time == 0:
             wait: float = 0
         else:
-            assert lastrun.completed is not None
-            wait = max(0, wait_time - (now - lastrun.completed).total_seconds())
+            assert last_run.completed is not None
+            if last_run.completed >= compile.requested:
+                wait = max(0, wait_time - (last_run.completed - compile.requested).total_seconds())
+            else:
+                wait = 0
         if wait > 0:
             LOGGER.info(
-                "Last recompile longer than %s ago (last was at %s), waiting for %d", wait_time, lastrun.completed, wait
+                "Last recompile less than %s seconds ago (last was at %s), waiting for %f seconds",
+                wait_time,
+                last_run.completed,
+                wait,
```

Comment:
This message is no longer accurate.
```suggestion
                "Compile requested less than %s seconds ago, waiting for %f seconds",
                wait_time,
                wait,
```